docs/RFCS/split_sql_syntax.md

```diff
@@ -0,0 +1,138 @@
+the replicas and range leaders of each split.
+
+The secondary motivation is for backup/restore, which needs to introduce a
+specific set of key splits. Currently this is done by custom code which works as
+follows: the keys are sorted; a split is introduced on the middle key, then
+the splits to the left and respectively to the right are processed recursively
+(in parallel). We want the new syntax to implement this algorithm so that
+backup/restore can use it. Note that in this case, we already have the split keys
+(as opposed to primary key values). In addition to backup/restore, there are
+more usecases for pre-spliting ranges (for benchmarks, importing, etc).
+
+We currently have the `ALTER TABLE/INDEX SPLIT AT` statement which allows
+introducing a table or index split at specific column values. The main drawback
+is that it cannot be used *programmatically*: each split must be in its own
+statement; we cannot generate (e.g. via a `SELECT`) a set of values where to
+split. We want to write tests like the one linked above without hardcoding each
+split in the test file; and we want to be able to easily change the test table
+sizes. In addition, the `SPLIT AT` statments don't support control of
+replication.
+
+
+# Detailed design
+
+Three new pieces of SQL syntax are proposed.
+
+### 1. `key_prefix` function ###
+
+The `key_prefix` function is a function that takes a table name (optionally with
+an index) and a set of column values for a prefix of the primary key (or index
+columns) and returns the corresponding key (as `BYTES`). Example:
+
+```sql
+CREATE TABLE t (
+  k1 INT,
+  k2 INT,
+  v INT,
+  w INT,
+  PRIMARY KEY(k1, k2),
+  INDEX vw (v, w)
+)
+
+SELECT key_prefix('t', 1, 2)     -- returns /t/primary/1/2
+SELECT key_prefix('db.t', 3)     -- returns /t/primary/3
+SELECT key_prefix('t@vw', 4, 5)  -- returns /t/vw/4/5
+SELECT key_prefix('t@vw', 1)     -- returns /t/vw/1
+```
+
+#### `pretty_key` function ####
+
+For debugging/testing purposes, we will also need a `pretty_key` function that
+takes a `BYTES` key generated by `key_prefix` and returns a pretty-printed string.
+
+### 2. `KVSPLIT` statement ###
+
+`KVSPLIT <select_clause>`
+
+The `KVSPLIT` statement takes a select clause that produces a single column of
+`BYTES` keys and executes the splits. Examples:
+
+```sql
+KVSPLIT VALUES ('..'), ('..')
+KVSPLIT SELECT key_prefix('t', 1, 2)
+KVSPLIT SELECT key_prefix('t', i*10) FROM GENERATE_SERIES(1, 5) AS x(i)
+```
+
+`KVSPLIT` is implemented as follows:
+ - the input keys are sorted;
+ - a split is performed of the middle key; recursively the left and right sets
+   are split in parallel;
+ - a queue or semaphore can be used to limit the amount of split operations that
+   happen at the same time.
+
+The order in which we perform splits matters for achieving parallelism: if we
+execute the splits in order, we can only do one split at a time because every
+split runs on an existing range.
```

Comment:
By this argument I think we should first split on the 2nd and next-to-last key first, and only then the middle keys in parallel, otherwise there would be many operations splitting on the sides where the ranges are touching already existing ranges.